File: product/tradercockpit/capability_registry.py
"""Backend capability/add-on registry.

One authority supplies typed descriptors. Add-ons bind registered slots only.
They cannot inject HTML/script, rewrite top-level navigation, or claim producer truth.
Unknown descriptor versions fail closed.
"""

from __future__ import annotations

from json import JSONDecodeError, loads
from pathlib import Path
from typing import Any

from tradercockpit.native_plugins import (
    NativePluginError,
    load_native_plugin_catalog,
    plugin_runtime_state,
    stage_native_plugin,
)


REGISTRY_SCHEMA = "tc.capability-addon-registry.v1"
ADDON_SCHEMA = "tc.capability-addon.v1"
ADDON_DESCRIPTOR_VERSION = 1
REGISTRY_API_PATH = "/api/capabilities"
ADDONS_DIR_NAME = "addons"
NONE_SCHEMA = "tc.capability-addon.none.v1"
NAV_AUTHORITY = "platform"

# Must match the browser's APP_SURFACES exactly: the registry payload is the nav
# authority the frontend validates against. Retester and Optimizer are native SQX
# module identities that redirect to Builder, not top-level surfaces.
PLATFORM_SURFACES = (
    "home", "builder", "data-manager", "custom-projects", "apollo", "operate", "settings",
)

REGISTERED_SLOTS = (
    {
        "id": "explore.extensions",
        "surface": "settings",
        "kind": "status_card",
        "label": "Settings plugin catalog",
    },
    {
        "id": "automation.extensions",
        "surface": "custom-projects",
        "kind": "status_card",
        "label": "Custom projects extensions",
    },
    {
        "id": "settings.extensions",
        "surface": "settings",
        "kind": "status_card",
        "label": "Settings extensions",
    },
)

_SLOT_IDS = frozenset(slot["id"] for slot in REGISTERED_SLOTS)
_SLOT_SURFACES = frozenset(slot["surface"] for slot in REGISTERED_SLOTS)
_ADDON_KEYS = frozenset(
    {
        "schema",
        "descriptor_version",
        "id",
        "version",
        "producer",
        "availability",
        "slot",
        "kind",
        "package",
        "native_placement",
        "source_url",
        "runtime",
        "config_schema",
        "read_schema",
        "action_schema",
        "presentation",
    }
)
_PRESENTATION_KEYS = frozenset({"title", "detail", "job", "opens_in", "controls"})
_CONTROL_KEYS = frozenset({"label", "detail"})
_PRODUCERS = frozenset({"operator", "platform", "native_sqx"})
_BUNDLED_PRODUCERS = frozenset({"native_sqx", "platform"})
_AVAILABILITY = frozenset({"ready", "unavailable"})
_KINDS = frozenset({"results_plugin", "authoring_skill", "sxp_extension", "operator"})
_ID_CHARS = frozenset("abcdefghijklmnopqrstuvwxyz0123456789.-")
_ALLOWED_CONTRACT_SCHEMAS = frozenset({NONE_SCHEMA})


class CapabilityRegistryError(ValueError):
    def __init__(self, code: str, detail: str) -> None:
        super().__init__(f"{code}: {detail}")
        self.code = code
        self.detail = detail


def registered_slots() -> list[dict[str, str]]:
    return [dict(slot) for slot in REGISTERED_SLOTS]


def _optional_home(sqx_home: Path | str | None) -> Path | None:
    if sqx_home is None:
        return None
    text = str(sqx_home).strip()
    return Path(text) if text else None


def capability_registry_record(
    data_root: Path | str | None = None,
    sqx_home: Path | str | None = None,
) -> dict[str, Any]:
    """Return the typed add-on registry, including packaged native SQX plugins."""

    addons: list[dict[str, Any]] = []
    refused: list[dict[str, str]] = []
    seen: set[str] = set()
    sqx_home = _optional_home(sqx_home)
    try:
        for entry in load_native_plugin_catalog():
            addons.append(_bundled_addon(entry, sqx_home, seen))
    except Exception as exc:  # noqa: BLE001 - catalog must fail closed, not crash the desktop
        refused.append({"source": "native_plugins/catalog.json", "reason_code": "native_plugin_catalog_invalid", "detail": str(exc)})

    try:
        directory = _addons_dir(data_root)
    except CapabilityRegistryError as exc:
        return _record(addons, refused + [{"source": ADDONS_DIR_NAME, "reason_code": exc.code, "detail": exc.detail}], status="unavailable")

    if directory is None:
        return _record(addons, refused)

    for path in sorted(directory.iterdir(), key=lambda item: item.name):
        if path.name.startswith(".") or path.suffix.lower() != ".json":
            continue
        try:
            addons.append(_load_addon_file(directory, path, seen))
            seen.add(addons[-1]["id"])
        except CapabilityRegistryError as exc:
            refused.append({"source": path.name, "reason_code": exc.code, "detail": exc.detail})
    return _record(addons, refused)


def extensions_status_record(
    data_root: Path | str | None = None,
    sqx_home: Path | str | None = None,
) -> dict[str, Any]:
    """Compact /api/status view of the same registry authority."""

    registry = capability_registry_record(data_root, sqx_home)
    return {
        "status": registry["status"],
        "reason_code": registry["reason_code"],
        "detail": registry["detail"],
        "registry_schema": REGISTRY_SCHEMA,
        "nav_authority": NAV_AUTHORITY,
        "slot_count": len(REGISTERED_SLOTS),
        "addon_count": registry["addon_count"],
        "refused_count": registry["refused_count"],
    }


def stage_addon(plugin_id: str, sqx_home: Path | str | None) -> dict[str, Any]:
    home = _optional_home(sqx_home)
    if home is None:
        raise CapabilityRegistryError(
            "runtime_not_configured",
            "Install requires a verified StrategyQuant X 144.2953 runtime. The browser cannot choose this path.",
        )
    try:
        return stage_native_plugin(plugin_id, home)
    except NativePluginError as exc:
        raise CapabilityRegistryError(exc.code, exc.detail) from exc


def _record(addons: list[dict[str, Any]], refused: list[dict[str, str]], *, status: str = "ready") -> dict[str, Any]:
    reason = None if status == "ready" else (refused[0]["reason_code"] if refused else "addon_store_unreadable")
    detail = (
        "Native StrategyQuant X plugins are packaged. Results plugins install into SQX; "
        "their settings stay in StrategyQuant X. Add-ons cannot rewrite top-level navigation "
        "or inject script/HTML."
        if status == "ready"
        else (refused[0]["detail"] if refused else "The add-on store could not be read.")
    )
    if status == "ready" and refused:
        detail = (
            f"{detail} {len(refused)} descriptor"
            f"{'' if len(refused) == 1 else 's'} failed closed and "
            "were not bound."
        )
    return {
        "schema": REGISTRY_SCHEMA,
        "status": status,
        "reason_code": reason,
        "detail": detail,
        "nav_authority": NAV_AUTHORITY,
        "surfaces": list(PLATFORM_SURFACES),
        "slots": registered_slots(),
        "addons": addons,
        "refused": refused,
        "addon_count": len(addons),
        "refused_count": len(refused),
    }


def _addons_dir(data_root: Path | str | None) -> Path | None:
    if data_root is None:
        return None
    root = Path(data_root).expanduser().resolve()
    candidate = root / ADDONS_DIR_NAME
    if not candidate.exists():
        return None
    if candidate.is_symlink() or candidate.is_junction():
        raise CapabilityRegistryError(
            "addon_store_path_escape",
            "Add-on store must be a real directory inside the application data root.",
        )
    if not candidate.is_dir():
        raise CapabilityRegistryError(
            "addon_store_unreadable",
            "Add-on store must be a directory of JSON descriptors.",
        )
    try:
        resolved = candidate.resolve()
        resolved.relative_to(root)
    except (OSError, RuntimeError, ValueError) as exc:
        raise CapabilityRegistryError(
            "addon_store_path_escape",
            "Add-on store resolved outside the application data root.",
        ) from exc
    return resolved


def _load_addon_file(directory: Path, path: Path, seen_ids: set[str]) -> dict[str, Any]:
    if path.is_symlink() or path.is_junction():
        raise CapabilityRegistryError("addon_path_escape", f"{path.name} is a symlink or junction and was refused.")
    try:
        resolved = path.resolve()
        resolved.relative_to(directory)
    except (OSError, RuntimeError, ValueError) as exc:
        raise CapabilityRegistryError("addon_path_escape", f"{path.name} resolved outside the add-on store.") from exc
    if not resolved.is_file():
        raise CapabilityRegistryError("addon_unreadable", f"{path.name} is not a descriptor file.")
    try:
        payload = loads(resolved.read_text(encoding="utf-8"))
    except (OSError, UnicodeDecodeError) as exc:
        raise CapabilityRegistryError("addon_unreadable", f"{path.name} could not be read.") from exc
    except JSONDecodeError as exc:
        raise CapabilityRegistryError("addon_descriptor_invalid", f"{path.name} is not JSON.") from exc
    return addon_from_payload(payload, seen_ids, allow_native=False)


def _bundled_addon(entry: dict[str, Any], sqx_home: Path | str | None, seen: set[str]) -> dict[str, Any]:
    runtime = plugin_runtime_state(entry, sqx_home)
    payload = {
        "schema": ADDON_SCHEMA,
        "descriptor_version": ADDON_DESCRIPTOR_VERSION,
        "id": entry["id"],
        "version": entry["version"],
        "producer": "native_sqx",
        "availability": "ready" if runtime["status"] != "unavailable" else "unavailable",
        "slot": entry["slot"],
        "kind": entry["kind"],
        "package": entry.get("package"),
        "native_placement": entry.get("native_placement"),
        "source_url": entry.get("source_url"),
        "runtime": runtime,
        "config_schema": NONE_SCHEMA,
        "read_schema": NONE_SCHEMA,
        "action_schema": None,
        "presentation": entry["presentation"],
    }
    addon = addon_from_payload(payload, seen, allow_native=True)
    seen.add(addon["id"])
    return addon


def addon_from_payload(payload: object, seen_ids: set[str] | None = None, *, allow_native: bool = False) -> dict[str, Any]:
    if not isinstance(payload, dict):
        raise CapabilityRegistryError("addon_descriptor_invalid", "Add-on descriptor must be an object.")
    extra = set(payload) - _ADDON_KEYS
    if extra:
        raise CapabilityRegistryError(
            "addon_descriptor_invalid",
            "Add-on descriptor has unsupported keys. Add-ons cannot rewrite navigation or inject markup.",
        )
    if payload.get("schema") != ADDON_SCHEMA:
        raise CapabilityRegistryError("addon_descriptor_invalid", "Add-on descriptor schema is not registered.")
    if payload.get("descriptor_version") != ADDON_DESCRIPTOR_VERSION:
        raise CapabilityRegistryError("addon_descriptor_version_unsupported", "Unknown add-on descriptor version failed closed.")

    identity = payload.get("id")
    version = payload.get("version")
    producer = payload.get("producer")
    availability = payload.get("availability")
    slot = payload.get("slot")
    kind = payload.get("kind", "operator")
    package = payload.get("package")
    native_placement = payload.get("native_placement")
    source_url = payload.get("source_url")
    runtime = payload.get("runtime")
    config_schema = payload.get("config_schema")
    read_schema = payload.get("read_schema")
    action_schema = payload.get("action_schema")
    presentation = payload.get("presentation")

    if not _addon_id(identity):
        raise CapabilityRegistryError("addon_identity_invalid", "Add-on identity is not a registered slug.")
    if seen_ids is not None and identity in seen_ids:
        raise CapabilityRegistryError("addon_identity_duplicate", f"Add-on identity {identity} is already bound.")
    if not isinstance(version, str) or not version or not _plain_text(version) or len(version) > 32:
        raise CapabilityRegistryError("addon_descriptor_invalid", "Add-on version is invalid.")
    if producer == "native_sqx" and not allow_native:
        raise CapabilityRegistryError(
            "addon_producer_refused",
            "Operator add-ons cannot claim native producer truth. Packaged SQX plugins come from the desktop catalog.",
        )
    if producer not in _PRODUCERS:
        raise CapabilityRegistryError(
            "addon_producer_refused",
            "Add-on producer must be operator, platform, or a packaged native SQX plugin.",
        )
    if availability not in _AVAILABILITY:
        raise CapabilityRegistryError("addon_descriptor_invalid", "Add-on availability is invalid.")
    if slot not in _SLOT_IDS:
        raise CapabilityRegistryError(
            "addon_slot_unregistered",
            "Add-on placement is not a registered typed slot. Top-level navigation cannot be extended.",
        )
    if kind not in _KINDS:
        raise CapabilityRegistryError("addon_descriptor_invalid", "Add-on kind is not registered.")
    if package is not None and (not isinstance(package, str) or not _plain_text(package) or "/" in package or "\\" in package or len(package) > 80):
        raise CapabilityRegistryError("addon_descriptor_invalid", "Add-on package name is invalid.")
    if native_placement is not None and (
        not isinstance(native_placement, str)
        or not _plain_text(native_placement)
        or native_placement.startswith("/")
        or ".." in Path(native_placement).parts
        or len(native_placement) > 160
    ):
        raise CapabilityRegistryError("addon_descriptor_invalid", "Native placement must be a relative runtime path.")
    if source_url is not None and (
        not isinstance(source_url, str)
        or not source_url.startswith("https://strategyquant.com/")
        or not _plain_text(source_url)
        or len(source_url) > 240
    ):
        raise CapabilityRegistryError("addon_descriptor_invalid", "Add-on source URL is not a StrategyQuant codebase URL.")
    if runtime is None:
        runtime = {
            "status": "packaged",
            "installed": False,
            "stageable": False,
            "detail": "Operator add-on. No native Results install.",
        }
    runtime = _runtime(runtime)
    if config_schema not in _ALLOWED_CONTRACT_SCHEMAS or read_schema not in _ALLOWED_CONTRACT_SCHEMAS:
        raise CapabilityRegistryError("addon_contract_invalid", "Add-on config/read schemas must be the registered none contract.")
    if action_schema is not None:
        raise CapabilityRegistryError(
            "addon_action_refused",
            "Add-ons have no mutation contract. Install uses the canonical capability API. Plugin settings stay in StrategyQuant X.",
        )
    return {
        "schema": ADDON_SCHEMA,
        "descriptor_version": ADDON_DESCRIPTOR_VERSION,
        "id": identity,
        "version": version,
        "producer": producer,
        "availability": availability,
        "slot": slot,
        "kind": kind,
        "package": package,
        "native_placement": native_placement,
        "source_url": source_url,
        "runtime": runtime,
        "config_schema": config_schema,
        "read_schema": read_schema,
        "action_schema": None,
        "presentation": _presentation(presentation),
    }


def _runtime(value: object) -> dict[str, Any]:
    if not isinstance(value, dict):
        raise CapabilityRegistryError("addon_descriptor_invalid", "Add-on runtime state must be an object.")
    status = value.get("status")
    if status not in {"packaged", "installed", "runtime_not_configured", "unavailable"}:
        raise CapabilityRegistryError("addon_descriptor_invalid", "Add-on runtime status is invalid.")
    detail = value.get("detail")
    if not isinstance(detail, str) or not _plain_text(detail) or len(detail) > 240:
        raise CapabilityRegistryError("addon_descriptor_invalid", "Add-on runtime detail must be plain text.")
    installed = value.get("installed") is True
    stageable = value.get("stageable") is True
    return {"status": status, "installed": installed, "stageable": stageable, "detail": detail}


def _presentation(value: object) -> dict[str, Any]:
    if not isinstance(value, dict):
        raise CapabilityRegistryError("addon_presentation_invalid", "Add-on presentation must be a typed object.")
    if set(value) - _PRESENTATION_KEYS:
        raise CapabilityRegistryError(
            "addon_presentation_invalid",
            "Add-on presentation may only include title, detail, job, opens_in, and controls. HTML/script are refused.",
        )
    title = value.get("title")
    detail = value.get("detail")
    job = value.get("job", "")
    opens_in = value.get("opens_in", "")
    if not isinstance(title, str) or not _plain_text(title) or not title.strip() or len(title) > 80:
        raise CapabilityRegistryError("addon_presentation_invalid", "Add-on title must be plain text.")
    if not isinstance(detail, str) or not _plain_text(detail) or not detail.strip() or len(detail) > 400:
        raise CapabilityRegistryError("addon_presentation_invalid", "Add-on detail must be plain text.")
    if job and (not isinstance(job, str) or not _plain_text(job) or len(job) > 160):
        raise CapabilityRegistryError("addon_presentation_invalid", "Add-on job must be plain text.")
    if opens_in and (not isinstance(opens_in, str) or not _plain_text(opens_in) or len(opens_in) > 80):
        raise CapabilityRegistryError("addon_presentation_invalid", "Add-on opens_in must be plain text.")
    controls = _controls(value.get("controls", []))
    return {
        "title": title.strip(),
        "detail": detail.strip(),
        "job": job.strip() if isinstance(job, str) else "",
        "opens_in": opens_in.strip() if isinstance(opens_in, str) else "",
        "controls": controls,
    }


def _controls(value: object) -> list[dict[str, str]]:
    if value in (None, []):
        return []
    if not isinstance(value, list) or len(value) > 6:
        raise CapabilityRegistryError("addon_presentation_invalid", "Add-on controls must be a short typed list.")
    controls: list[dict[str, str]] = []
    for item in value:
        if not isinstance(item, dict) or set(item) - _CONTROL_KEYS:
            raise CapabilityRegistryError("addon_presentation_invalid", "Each control may only include label and detail.")
        label = item.get("label")
        detail = item.get("detail")
        if not isinstance(label, str) or not _plain_text(label) or not label.strip() or len(label) > 40:
            raise CapabilityRegistryError("addon_presentation_invalid", "Control label must be plain text.")
        if not isinstance(detail, str) or not _plain_text(detail) or not detail.strip() or len(detail) > 160:
            raise CapabilityRegistryError("addon_presentation_invalid", "Control detail must be plain text.")
        controls.append({"label": label.strip(), "detail": detail.strip()})
    return controls



def _addon_id(value: object) -> bool:
    if not isinstance(value, str) or len(value) < 3 or len(value) > 80:
        return False
    if value[0] not in "abcdefghijklmnopqrstuvwxyz":
        return False
    if any(char not in _ID_CHARS for char in value):
        return False
    if ".." in value or value.startswith(".") or value.endswith("."):
        return False
    return "." in value


def _plain_text(value: str) -> bool:
    if any(ord(char) < 32 for char in value):
        return False
    lowered = value.lower()
    if "<" in value or ">" in value or "</" in value:
        return False
    if "javascript:" in lowered or "data:text/html" in lowered:
        return False
    return True


def slot_surface(slot_id: str) -> str | None:
    if slot_id not in _SLOT_IDS:
        return None
    return next(slot["surface"] for slot in REGISTERED_SLOTS if slot["id"] == slot_id)


assert "home" in PLATFORM_SURFACES
assert all(slot["surface"] in _SLOT_SURFACES for slot in REGISTERED_SLOTS)
assert not any(slot["kind"] == "navigation" for slot in REGISTERED_SLOTS)
assert set(PLATFORM_SURFACES) == {
    "home",
    "builder",
    "data-manager",
    "custom-projects",
    "apollo",
    "operate",
    "settings",
}

File: web/capability-registry.mjs
// Typed add-on registry. The backend is the only catalog. The browser paints
// packaged native SQX plugins into registered slots and never rewrites navigation.

import { APP_SURFACES } from "./model.mjs";
import { actionButton, escapeHtml, readable, unavailable } from "./ui.mjs";

export const CAPABILITY_REGISTRY_SCHEMA = "tc.capability-addon-registry.v1";
export const CAPABILITY_ADDON_SCHEMA = "tc.capability-addon.v1";
export const CAPABILITY_REGISTRY_API_PATH = "/api/capabilities";
export const ADDON_DESCRIPTOR_VERSION = 1;
export const NONE_SCHEMA = "tc.capability-addon.none.v1";
export const REGISTERED_SLOT_IDS = Object.freeze([
  "explore.extensions",
  "automation.extensions",
  "settings.extensions",
]);
export const CAPABILITY_VIEWS = Object.freeze(["catalog", "results", "install"]);

const PLATFORM_SURFACE_IDS = Object.freeze(APP_SURFACES.map((surface) => surface.id));
const REGISTRY_KEYS = Object.freeze([
  "schema",
  "status",
  "reason_code",
  "detail",
  "nav_authority",
  "surfaces",
  "slots",
  "addons",
  "refused",
  "addon_count",
  "refused_count",
]);
const ADDON_KEYS = Object.freeze([
  "schema",
  "descriptor_version",
  "id",
  "version",
  "producer",
  "availability",
  "slot",
  "kind",
  "package",
  "native_placement",
  "source_url",
  "runtime",
  "config_schema",
  "read_schema",
  "action_schema",
  "presentation",
]);
const SLOT_KEYS = Object.freeze(["id", "surface", "kind", "label"]);
const PRESENTATION_KEYS = Object.freeze(["title", "detail", "job", "opens_in", "controls"]);
const CONTROL_KEYS = Object.freeze(["label", "detail"]);
const RUNTIME_KEYS = Object.freeze(["status", "installed", "stageable", "detail"]);
const PRODUCERS = Object.freeze(["operator", "platform", "native_sqx"]);
const KINDS = Object.freeze(["results_plugin", "authoring_skill", "sxp_extension", "operator"]);
const RUNTIME_STATUSES = Object.freeze(["packaged", "installed", "runtime_not_configured", "unavailable"]);

function object(value) {
  return value && typeof value === "object" && !Array.isArray(value) ? value : null;
}

function exactKeys(record, allowed) {
  const keys = Object.keys(record);
  return keys.length === allowed.length && keys.every((key) => allowed.includes(key));
}

function markupFree(...parts) {
  return !parts.some((part) => typeof part === "string" && /[<>]/.test(part));
}

function runtimeBadge(runtime) {
  if (runtime.installed) return { label: "Installed in SQX", tone: "ready" };
  if (runtime.status === "packaged") return { label: "Packaged", tone: "ready" };
  if (runtime.status === "runtime_not_configured") return { label: "Needs SQX runtime", tone: "pending" };
  return { label: readable(runtime.status), tone: "unavailable" };
}

function kindLabel(kind) {
  if (kind === "results_plugin") return "SQX Results plugin";
  if (kind === "sxp_extension") return "SQX Results extension";
  if (kind === "authoring_skill") return "Native authoring";
  return "Operator add-on";
}

export function capabilityRegistryFromPayload(payload) {
  const record = object(payload);
  if (
    !record
    || !exactKeys(record, REGISTRY_KEYS)
    || record.schema !== CAPABILITY_REGISTRY_SCHEMA
    || record.nav_authority !== "platform"
    || !["ready", "unavailable"].includes(record.status)
    || (record.reason_code !== null && (typeof record.reason_code !== "string" || !record.reason_code))
    || typeof record.detail !== "string"
    || !Array.isArray(record.surfaces)
    || record.surfaces.length !== PLATFORM_SURFACE_IDS.length
    || record.surfaces.some((id, index) => id !== PLATFORM_SURFACE_IDS[index])
    || !Array.isArray(record.slots)
    || record.slots.length !== REGISTERED_SLOT_IDS.length
    || !Array.isArray(record.addons)
    || !Array.isArray(record.refused)
    || record.addon_count !== record.addons.length
    || record.refused_count !== record.refused.length
  ) {
    throw new Error("Capability registry schema mismatch");
  }

  const slotIds = [];
  for (const slot of record.slots) {
    if (
      !object(slot)
      || !exactKeys(slot, SLOT_KEYS)
      || typeof slot.id !== "string"
      || !REGISTERED_SLOT_IDS.includes(slot.id)
      || slotIds.includes(slot.id)
      || typeof slot.surface !== "string"
      || !PLATFORM_SURFACE_IDS.includes(slot.surface)
      || slot.kind !== "status_card"
      || typeof slot.label !== "string"
    ) {
      throw new Error("Capability registry slot is not registered");
    }
    slotIds.push(slot.id);
  }
  if (REGISTERED_SLOT_IDS.some((id) => !slotIds.includes(id))) {
    throw new Error("Capability registry is missing a registered slot");
  }

  const addons = record.addons.map((item) => addonFromPayload(item, slotIds));
  const seen = new Set();
  for (const addon of addons) {
    if (seen.has(addon.id)) throw new Error("Capability registry add-on identity is duplicated");
    seen.add(addon.id);
  }
  for (const item of record.refused) {
    if (
      !object(item)
      || typeof item.source !== "string"
      || typeof item.reason_code !== "string"
      || typeof item.detail !== "string"
    ) {
      throw new Error("Capability registry refusal is invalid");
    }
  }
  return { ...record, addons };
}

export function addonFromPayload(value, slotIds = REGISTERED_SLOT_IDS) {
  const addon = object(value);
  const presentation = object(addon?.presentation);
  const runtime = object(addon?.runtime);
  const controls = Array.isArray(presentation?.controls) ? presentation.controls : null;
  if (
    !addon
    || !exactKeys(addon, ADDON_KEYS)
    || addon.schema !== CAPABILITY_ADDON_SCHEMA
    || addon.descriptor_version !== ADDON_DESCRIPTOR_VERSION
    || typeof addon.id !== "string"
    || typeof addon.version !== "string"
    || !PRODUCERS.includes(addon.producer)
    || (addon.availability !== "ready" && addon.availability !== "unavailable")
    || !slotIds.includes(addon.slot)
    || !KINDS.includes(addon.kind)
    || (addon.package !== null && typeof addon.package !== "string")
    || (addon.native_placement !== null && typeof addon.native_placement !== "string")
    || (addon.source_url !== null && typeof addon.source_url !== "string")
    || !runtime
    || !exactKeys(runtime, RUNTIME_KEYS)
    || !RUNTIME_STATUSES.includes(runtime.status)
    || typeof runtime.installed !== "boolean"
    || typeof runtime.stageable !== "boolean"
    || typeof runtime.detail !== "string"
    || addon.config_schema !== NONE_SCHEMA
    || addon.read_schema !== NONE_SCHEMA
    || addon.action_schema !== null
    || !presentation
    || !exactKeys(presentation, PRESENTATION_KEYS)
    || typeof presentation.title !== "string"
    || typeof presentation.detail !== "string"
    || typeof presentation.job !== "string"
    || typeof presentation.opens_in !== "string"
    || !controls
    || controls.length > 6
  ) {
    throw new Error("Add-on descriptor is invalid");
  }
  for (const control of controls) {
    if (
      !object(control)
      || !exactKeys(control, CONTROL_KEYS)
      || typeof control.label !== "string"
      || typeof control.detail !== "string"
      || !markupFree(control.label, control.detail)
    ) {
      throw new Error("Add-on presentation is invalid");
    }
  }
  if (!markupFree(
    addon.presentation.title,
    addon.presentation.detail,
    addon.presentation.job,
    addon.presentation.opens_in,
    addon.id,
    addon.version,
  )) {
    throw new Error("Add-on markup is refused");
  }
  return addon;
}

export async function fetchCapabilityRegistry(fetchImpl = globalThis.fetch) {
  if (typeof fetchImpl !== "function") throw new Error("capability registry fetch is unavailable");
  const response = await fetchImpl(CAPABILITY_REGISTRY_API_PATH, { headers: { accept: "application/json" } });
  if (!response?.ok) throw new Error(`capability registry request failed: ${response?.status ?? "unknown"}`);
  return capabilityRegistryFromPayload(await response.json());
}

export function selectCapabilityAddons(addons, view, slotId) {
  if (view === "catalog" || view === "install") return addons;
  if (view === "results") {
    return addons.filter((addon) => addon.kind === "results_plugin" || addon.kind === "sxp_extension");
  }
  return addons.filter((addon) => addon.slot === slotId);
}

function renderControls(controls) {
  if (!controls.length) return "";
  const rows = controls.map((control) => `<div class="plugin-control"><strong>${escapeHtml(control.label)}</strong><span>${escapeHtml(control.detail)}</span></div>`).join("");
  return `<div class="plugin-controls"><span class="plugin-controls-label">Adjust in StrategyQuant X</span>${rows}</div>`;
}

function renderStageAction(addon) {
  if (addon.kind === "authoring_skill") {
    return `<p class="note">Packaged for native authoring. It is not a Results tab.</p>`;
  }
  if (addon.runtime.installed) {
    return actionButton("Installed in SQX", {
      disabled: true,
      title: addon.runtime.detail,
    });
  }
  if (addon.runtime.stageable) {
    return actionButton("Install into SQX", {
      primary: true,
      attrs: `data-capability-stage="${escapeHtml(addon.id)}"`,
      title: "Copies this plugin into the authorized StrategyQuant X runtime. Settings stay in SQX Results.",
    });
  }
  return actionButton("Install into SQX", {
    disabled: true,
    attrs: `data-capability-stage="${escapeHtml(addon.id)}"`,
    title: addon.runtime.detail || "Install requires a verified StrategyQuant X runtime. The browser cannot choose this path.",
  });
}

function renderPluginCard(addon) {
  const badge = runtimeBadge(addon.runtime);
  const job = addon.presentation.job || addon.presentation.detail;
  const opens = addon.presentation.opens_in
    ? `<p class="plugin-opens">Opens in ${escapeHtml(addon.presentation.opens_in)}</p>`
    : "";
  return `<article class="plugin-card" data-capability-addon="${escapeHtml(addon.id)}" data-capability-kind="${escapeHtml(addon.kind)}"><header class="plugin-card-head"><div><span class="plugin-kind">${escapeHtml(kindLabel(addon.kind))}</span><h3>${escapeHtml(addon.presentation.title)}</h3></div><span class="status-badge status-${badge.tone}"><span class="status-dot"></span>${escapeHtml(badge.label)}</span></header><p class="plugin-job">${escapeHtml(job)}</p>${opens}${renderControls(addon.presentation.controls)}<footer class="plugin-card-foot">${renderStageAction(addon)}</footer></article>`;
}

function renderInstallRow(addon) {
  const badge = runtimeBadge(addon.runtime);
  return `<div class="plugin-install-row" data-capability-addon="${escapeHtml(addon.id)}"><div><strong>${escapeHtml(addon.presentation.title)}</strong><p>${escapeHtml(addon.runtime.detail || addon.presentation.job || addon.presentation.detail)}</p></div><div class="plugin-install-actions"><span class="status-badge status-${badge.tone}"><span class="status-dot"></span>${escapeHtml(badge.label)}</span>${renderStageAction(addon)}</div></div>`;
}

export function renderCapabilitySlot(registry, slotId, view = "slot") {
  const manifest = capabilityRegistryFromPayload(registry);
  const slot = manifest.slots.find((item) => item.id === slotId);
  if (!slot) throw new Error("Capability slot is not registered");
  const bound = selectCapabilityAddons(manifest.addons, view, slotId);
  const layout = view === "install" ? "install" : "catalog";
  const rows = bound.length
    ? (layout === "install"
      ? `<div class="plugin-install-list">${bound.map(renderInstallRow).join("")}</div>`
      : `<div class="plugin-shelf">${bound.map(renderPluginCard).join("")}</div>`)
    : unavailable(
      "No native plugins in this view",
      "Packaged StrategyQuant X plugins bind registered slots only. They cannot invent a placement or rewrite navigation.",
      { compact: true },
    );
  const refused = manifest.refused.length
    ? `<p class="note">${escapeHtml(String(manifest.refused.length))} descriptor${manifest.refused.length === 1 ? "" : "s"} failed closed and ${manifest.refused.length === 1 ? "was" : "were"} not bound.</p>`
    : "";
  return `<section data-capability-slot-body="${escapeHtml(slotId)}" data-capability-view-body="${escapeHtml(view)}">${rows}${refused}</section>`;
}

function registryHosts() {
  return [...document.querySelectorAll("[data-capability-registry][data-capability-slot]")];
}

let generation = 0;

async function paintRegistry() {
  const hosts = registryHosts();
  if (!hosts.length) return;
  if (hosts.every((host) => host.getAttribute("data-capability-registry-state"))) return;
  const current = ++generation;
  try {
    const registry = await fetchCapabilityRegistry();
    if (current !== generation) return;
    for (const host of hosts) {
      if (!host.isConnected) continue;
      const slotId = host.getAttribute("data-capability-slot") || "";
      const view = host.getAttribute("data-capability-view") || "slot";
      try {
        host.setAttribute("data-capability-registry-state", "ready");
        host.innerHTML = renderCapabilitySlot(registry, slotId, view);
      } catch {
        host.setAttribute("data-capability-registry-state", "unavailable");
        host.innerHTML = unavailable(
          "Typed add-on slot unavailable",
          "This placement is not a registered slot. Add-ons cannot rewrite navigation.",
          { tone: "error", compact: true },
        );
      }
    }
  } catch (error) {
    if (current !== generation) return;
    const detail = error instanceof Error ? error.message : "Capability registry unavailable";
    for (const host of hosts) {
      if (!host.isConnected) continue;
      host.setAttribute("data-capability-registry-state", "unavailable");
      host.innerHTML = unavailable("Native plugins unavailable", detail, { tone: "error", compact: true });
    }
  }
}

function resetRegistryPaint() {
  for (const host of registryHosts()) host.removeAttribute("data-capability-registry-state");
  generation += 1;
  void paintRegistry();
}

async function stagePlugin(pluginId) {
  const response = await fetch(CAPABILITY_REGISTRY_API_PATH, {
    method: "POST",
    headers: { accept: "application/json", "content-type": "application/json" },
    body: JSON.stringify({ action: "stage", id: pluginId }),
  });
  const payload = await response.json().catch(() => ({}));
  if (!response.ok) {
    throw new Error(typeof payload.detail === "string" ? payload.detail : `Install failed (${response.status})`);
  }
}

if (typeof document !== "undefined") {
  document.addEventListener("click", (event) => {
    const button = event.target instanceof Element ? event.target.closest("[data-capability-stage]") : null;
    if (!(button instanceof HTMLButtonElement) || button.disabled) return;
    event.preventDefault();
    const pluginId = button.getAttribute("data-capability-stage") || "";
    if (!pluginId) return;
    button.disabled = true;
    void stagePlugin(pluginId)
      .catch((error) => {
        const detail = error instanceof Error ? error.message : "Install failed";
        button.title = detail;
      })
      .finally(() => {
        resetRegistryPaint();
      });
  });
  const observer = new MutationObserver(() => {
    if (registryHosts().length) void paintRegistry();
    else generation += 1;
  });
  observer.observe(document.documentElement, { childList: true, subtree: true });
  void paintRegistry();
}
